test/Tester/StorageFacet/Feature.Abstractions/IStorageFeature.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System.Threading.Tasks;
+
+namespace Tester.StorageFacet.Abstractions
+{
+    /// <summary>
+    /// Primary storage feature interface.
+    ///  This is the actual functionality the users need.
+    /// </summary>
+    /// <typeparam name="TState"></typeparam>
+    public interface IStorageFeature<TState>
+    {
+        TState State { get; set; }
+
+        Task Save();
+
+        // Test calls - used to verify facet wiring works
+        string Name { get; }
+        string GetExtendedInfo();
+    }
+
+    /// <summary>
+    /// Feature configuration information which application layer can provide to the
+    ///  feature per instance (by grain type if using attributes).
+    /// </summary>
+    public interface IStorageFeatureConfig
+    {
+        string StateName { get; }
+    }
+
+    /// <summary>
+    /// Feature configuration utility class
+    /// </summary>
+    public class StorageFeatureConfig : IStorageFeatureConfig
+    {
+        public StorageFeatureConfig(string stateName)
+        {
+            this.StateName = stateName;
+        }
+
+        public string StateName { get; }
+    }
+
+    /// <summary>
+    /// Creates a storage feature from a configuration
+    /// </summary>
+    public interface IStorageFeatureFactory
+    {
+        IStorageFeature<TState> Create<TState>(IStorageFeatureConfig config);
```

Comment:
I prototyped both patterns, and this one worked better with DI and end user experience.

Issues I encountered:

- Having a typed factory would prohibit (or significantly complicate) managing  IStorageFeatureFactorys as named services.
- By having typed factories, each specialization would necessitate the construction of a new factory.
- Users that wished to have the factory injected in their grain constructor directly would need to have a factory for each type of data they wished to construct.

Regarding the MakeGenericMethod, it is only called when creating the argument factory the first time, all subsequent calls simply invoke the method.